MicroMacro/Assets/Scripts/Module/Scaling/ScaleStackSensor.cs

```diff
@@ -0,0 +1,195 @@
+        }
+        
+        private void OnScaleCompleted(ScaleEventArgs args)
+        {
+            // サイズ変動したらRayLengthを再計算
+            CalcRayLength();
+        }
+        
+        private void OnDestroy()
+        {
+            refScaler.OnScaleStarted -= OnScaleStarted;
+            scaler.OnScaleCompleted -= OnScaleCompleted;
+        }
+
+        /// <summary>
+        /// 飛ばすRayの長さを計算するクラス。Triggerの中心から半径＋プレイヤーの幅 
+        /// </summary>
+        private void CalcRayLength()
+        {
+            // 毎回bounds.extentsを取得するとRayの長さがずれるので、元サイズにlossyScale（ワールド座標）を掛けてスケール後の面の位置を再計算
+            // なぜこうなるのかあんまり納得いってない
+            Vector2 worldSize = Vector2.Scale(baseTriggerSize, transform.lossyScale);
+            
+            // プレイヤーサイズ分足す
+            rayLength.x = worldSize.x + playerSize.x;
+            rayLength.y = worldSize.y + playerSize.y;
+        }
+        
+        private void OnTriggerStay(Collider other)
+        {
+            if (other.CompareTag("Player"))
+            {
+                // トリガーからプレイヤーへの方向ベクトル
+                Vector2 trig2Player = (playerCol.transform.position - transform.position).normalized;
+                
+                rayDirection = Normalize4Direction(trig2Player);
+                ShootRay(rayDirection);
+            }
+        }
+        
+        private void OnTriggerExit(Collider other)
+        {
+            // リセット
+            if (other.CompareTag("Player"))
+            {
+                isStack = false;
+                rayDirection = Vector2.zero; 
+            }
+        }
+
+        /// <summary>
+        /// ベクトルを絶対値で比較し上下左右で一番近いdirectionを返す
+        /// 例: (0.5, 1) -> (0, 1)    
+        /// </summary>
+        const float HorizontalThreshold = 0.3f; // 閾値調整(小さいほど上下の判定エリアが狭くなるイメージ), プレイヤーサイズから計算してもいい 
+        private Vector2 Normalize4Direction(Vector2 direction)
+        {
+            float absX = Mathf.Abs(direction.x);
+            float absY = Mathf.Abs(direction.y);
+            if (absX > absY)
+            {
+                if (direction.x > 0)
+                    return Vector2.right; // 右
+                else
+                    return Vector2.left;  // 左
+            }
+
+            if (absX < absY)   
+            {
+                // 水平方向のずれが小さい時 (ほぼ真上か真下) は上下判定
+                if (absX < HorizontalThreshold)
+                {
+                    if (direction.y > 0)
+                        return Vector2.up;   // 上
+                    else
+                        return Vector2.down; // 下
+                }
+                else
+                {
+                    // 水平方向のずれが大きい時は左右判定
+                    if (direction.x > 0)
+                        return Vector2.right; 
+                    else
+                        return Vector2.left; 
+                }
+            }
+
+            // 角の時はゼロベクトル返す(ほぼ起きない)
+            Debug.LogWarning($"無効なdirectionが渡されました: {direction}");
+            return Vector2.zero;
+        }
```

Comment:
条件が分散してるのでまとめた方が個人的には良いと思う。
あとMathf.Signを使うともう少し簡潔に書ける。

```suggestion
        /// <summary>
        /// ベクトルを絶対値で比較し上下左右で一番近いdirectionを返す
        /// 例: (0.5, 1) -> (0, 1)    
        /// </summary>
        const float HorizontalThreshold = 0.3f; // 閾値調整(小さいほど上下の判定エリアが狭くなるイメージ), プレイヤーサイズから計算してもいい 
        private Vector2 Normalize4Direction(Vector2 direction)
        {
            // ゼロベクトルの場合は処理しない
            if (direction == Vector2.zero) 
                return Vector2.zero;
        
            float absX = Mathf.Abs(direction.x);
            float absY = Mathf.Abs(direction.y);
        
            // 上下方向と判定する条件
            // Y成分がX成分より大きく、かつX成分が閾値より小さい場合
            // 斜めベクトル(absY == absX)の場合は、左右ベクトルとして処理される
            if (absY > absX && absX < HorizontalThreshold)
            {
                // 上下方向
                return new Vector2(0, Mathf.Sign(direction.y));
            }
            else
            {
                // それ以外はすべて左右方向
                return new Vector2(Mathf.Sign(direction.x), 0);
            }
        }
```